astroid/inference_tip.py

```diff
@@ -8,13 +8,21 @@
 
 import wrapt
 
-# pylint: disable=dangerous-default-value
 from astroid.exceptions import InferenceOverwriteError
 from astroid.nodes import NodeNG
 
+InferFn = typing.Callable[..., typing.Any]
+
+_cache: typing.Dict[typing.Tuple[InferFn, NodeNG], typing.Any] = {}
+
+
+def clear_inference_tip_cache():
```

Comment:
Alternatively, I can add a single `clear_caches()` entry point which also clears LRU caches on various methods in `astroid` (as discussed in #792). Wdyt?

Reply:
What do you think about using the `@lru_cache(maxsize=?)` decorator to handle that ? I think it would remove the need for the API. But maybe we'd need a mechanism to set the max size instead ?

Reply:
We'd need some extra logic to account for the return value being a generator, right?

Otherwise, `lru_cache` would do the job. I'd still prefer for us to have an explicit way to clear the cache, because regardless of capacity, references to nodes will keep (potentially expensive) object around in memory for longer.

Reply:
I don't know lru_cache well I did not know of the potential problem with generator. Just thought that we might as well use a specialized library to have a sane limit without any human intervention (Maybe 8 Go ?). (We can also have the API on top of it by the way).

Reply:
Sorry if I was unclear. I was referring to the bit of logic in the existing implementation, where the cached function does `itertools.tee` to make the cached generator "re-usable". We could, of course, eagerly materialize the generator into a `list`/`tuple` and cached that, instead. Does that sg?

Reply:
I don't understand the ``itertools.tee`` implementation well enough to guess if it will be better one way or the other. A way to test the speed of an implementation is to use the benchmark test in pylint by doing a local install of astroid and comparing the result. But I think as long as we set a limit to the cache we store it will probably be better than the current situation.

Reply:
I think `itertools.tee` has to cache the elements somewhere, and logically if one of the resulting iterators is fully consumed, `itertools.tee` will have to keep a full copy of the original iterable around internally. With this in mind, it seems we can just drop the `itertools.tee` and always return a list iterator.

As for `lru_cache`, I'm afraid we cannot use it here because the current implementation only uses the first argument as key (because the other one is `context`?), and `lru_cache` does not allow that.